dali/pipeline/operators/reader/loader/coco_loader.cc

```diff
@@ -12,433 +12,329 @@
+  }
 };
 
-bool LookaheadParser::EnterObject() {
-  if (st_ != kEnteringObject) {
-    st_  = kError;
-    return false;
-  }
+template<typename T>
+void dump_meta_file(std::vector<T> &input, const std::string path) {
+  std::ofstream file(path, std::ios_base::binary | std::ios_base::out);
+  DALI_ENFORCE(file, "CocoReader meta file error while saving: " + path);
 
-  ParseNext();
-  return true;
+  unsigned size = input.size();
+  file.write(reinterpret_cast<char*>(&size), sizeof(unsigned));
+  file.write(reinterpret_cast<char*>(input.data()), size * sizeof(T));
 }
 
-bool LookaheadParser::EnterArray() {
-  if (st_ != kEnteringArray) {
-    st_  = kError;
-    return false;
+void dump_filenames(const ImageIdPairs &image_id_pairs, const std::string path) {
+  std::ofstream file(path);
+  DALI_ENFORCE(file, "CocoReader meta file error while saving: " + path);
+  for (const auto &p : image_id_pairs) {
+    file << p.first << std::endl;
   }
-
-  ParseNext();
-  return true;
 }
 
-const char* LookaheadParser::NextObjectKey() {
-  if (st_ == kHasKey) {
-    const char* result = v_.GetString();
-    ParseNext();
-    return result;
-  }
-
-  if (st_ != kExitingObject) {
-    st_ = kError;
-    return 0;
-  }
+template<typename T>
+void load_meta_file(std::vector<T> &output, const std::string path) {
+  std::ifstream file(path);
+  DALI_ENFORCE(file, "CocoReader meta file error while loading for path: " + path);
 
-  ParseNext();
-  return 0;
+  unsigned size;
+  file.read(reinterpret_cast<char*>(&size), sizeof(unsigned));
+  output.resize(size);
+  file.read(reinterpret_cast<char*>(output.data()), size * sizeof(T));
 }
 
-bool LookaheadParser::NextArrayValue() {
-  if (st_ == kExitingArray) {
-    ParseNext();
-    return false;
-  }
 
-  if (st_ == kError || st_ == kExitingObject || st_ == kHasKey) {
-    st_ = kError;
-    return false;
-  }
-
-  return true;
-}
+void load_filenames(ImageIdPairs &image_id_pairs, const std::string path) {
+  std::ifstream file(path);
+  DALI_ENFORCE(file, "CocoReader meta file error while loading for path: " + path);
 
-int LookaheadParser::GetInt() {
-  if (st_ != kHasNumber || !v_.IsInt()) {
-    st_ = kError;
-    return 0;
+  int id = 0;
+  std::string filename;
+  while (file >> filename) {
+    image_id_pairs.emplace_back(std::move(filename), int{id});
+    ++id;
   }
-
-  int result = v_.GetInt();
-  ParseNext();
-  return result;
 }
 
-double LookaheadParser::GetDouble() {
-  if (st_ != kHasNumber) {
-    st_  = kError;
-    return 0.;
+void parse_image_infos(LookaheadParser &parser, std::vector<ImageInfo> &image_infos) {
+  RAPIDJSON_ASSERT(parser.PeekType() == kArrayType);
+  parser.EnterArray();
+  while (parser.NextArrayValue()) {
+    if (parser.PeekType() != kObjectType) {
+      continue;
+    }
+    parser.EnterObject();
+    ImageInfo image_info;
+    while (const char* internal_key = parser.NextObjectKey()) {
+      if (0 == strncmp(internal_key, "id", 2)) {
```

Comment:
How about some compile time strlen instead of hand written values.
You can even have something like (if I did it correctly):
```
template<size_t N>
inline int safe_strcmp(const char * str1, const char &str2[N] {
    return strncmp(str1, &str2, N-1)
}
```

Reply:
Done